File: Matlab/SFIRTI.m
%% SFIRTI (Sequential FIR Filter Type I) %%
%
% Description: 
% 1. Customazible and optimized Sequential FIR Filter.
% 2. The order of the filter has to be even (odd number of coefficients) 
% and its coefficients symmetric.
% 3. Customizable: The widths of input, output, and internal signals are 
% configurable, making natual growth and truncation of data possible. 
% 4. Optimized: As filter coefficients are symmetric, only half of values 
% that are symmetric to the middle address of the shift register are summed.
%
% Additional notes:
% 1. Stimuli is the result of adding 3 tones.
% 2. Stimuli, filter response, and filter coefficientes are exported to a
% txt file.
%
%
% Descripcion:
% 1. Filtro FIR secuencial personalizable.
% 2. El orden del filtro tiene que ser par (cantidad impar de coeficientes)
% y sus coeficientes simetricos.
% 3. Personalizable: El ancho de la senyal de entrada, salida, e internas son 
% configurables, haciendo posible el truncado y el crecimiento natural de los 
% datos internos.
% 4. Optimizado: Como el filtro es simetrico solo se realiza una suma de los 
% valores que son simetricos respecto a la direccion central del registro de 
% desplazamiento.
% 
% Notas adicionales:
% 1. La senyal de estimulo es la superposicion de 3 tonos.
% 2. Se exportan a txt la senyal de estimulo, la respuesta del filtro, y los
% valores de los coeficientes del filtro.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Configurable data %%
t_sim = 0.1;                    % Simulation time (seconds)

filter_order = 50;              % Filter order
filter_type = 1;                % 1: Low Pass Filter; 2: High Pass Filter; 3: Bandpass; 4: Bandstop
window = 1;                     % 1: hann; 2: hamming; 3: blackman; 4: triangular; 5: flat top; 6: nut tall
fc1 = 2000;                     % Cut-off frequency 1
fc2 = 3000;                     % Cut-off frequency 2. Only for highpass and bandstop filters
fs = 8000;                      % Sample frequency
f1 = 300;                       % Tone 1
f2 = 1000;                      % Tone 2
f3 = 2400;                      % Tone 3
amplitude = 2;                  % Amplitude of stimuli

wordlength_coeff = 10;          % Word length filter coefficients
fractionallength_coeff = 9;     % Fractional length filter coefficients

wordlength_input = 12;          % Word length stimuli
fractionallength_input = 10;    % Fractional length stimuli
    
wordlength_sum = 12;            % Word length adder
fractionallength_sum = 9;       % Fractional length adder

wordlength_mult = 12;           % Word length multiplier
fractionallength_mult = 8;      % Fractional length multiplier
        
wordlength_acc = 18;            % Word length accumulator
fractionallength_acc = 8;       % Fractional length accumulator

wordlength_output = 12;         % Word length output signal
fractionallength_output = 3;    % Fractional length output signal


%% Senyales intermedias %%
Ts = 0:1/fs:t_sim-1/fs;
signal1 = sin(2*pi*f1*Ts);
signal2 = sin(2*pi*f2*Ts);
signal3 = sin(2*pi*f3*Ts);
signal = (amplitude/3)*(signal1 + signal2 + signal3);
q = quantizer('fixed','round','saturate',[wordlength_input fractionallength_input]);
signal_q = quantize(q,signal);


%% Validacion de las entradas %%
if mod(filter_order,2) ~= 0
    error('Filter order has to be even')
elseif filter_type ~= 1 && filter_type ~= 2 && filter_type ~= 3 && filter_type ~= 4
    error('Invalid filter type.')
elseif window ~= 1 && window ~= 2 && window ~= 3 && window ~= 4 && window ~= 5 && window ~= 6
    error('Invalid window type.')
elseif fractionallength_coeff >= wordlength_coeff
    error('Coefficients length has to be larger than its fractional length')
elseif fractionallength_input >= wordlength_input
    error('Input length has to be larger than its fractional length')
elseif fractionallength_sum >= wordlength_sum
    error('Sum length has to be larger than its fractional length')
elseif fractionallength_mult >= wordlength_mult
    error('Multiplication length has to be larger than its fractional length')
elseif fractionallength_acc >= wordlength_acc
    error('Accumulator length has to be larger than its fractional length')
elseif fractionallength_output >= wordlength_output
    error('Output length has to be larger than its fractional length')
elseif fs/2 < fc1
    error('Cutt-off frequency cannot be higher than Nyquist frequency') 
elseif filter_type == 3 || filter_type == 4
    if fc2 <= fc1
        error('Cut-off frequency has to be increasing')
    end 
    if fs/2 < fc2
        error('Cutt-off frequency cannot be higher than Nyquist frequency')
    end 
end


%% Filtro %%
if filter_type == 3 || filter_type == 4 
    fc = [fc1 fc2]/(fs/2);
else
    fc = fc1/(fs/2);
end

type = ["low" "high" "bandpass" "stop"];
filter_type = char(type(filter_type));

window_type = {hann(filter_order+1) hamming(filter_order+1) blackman(filter_order+1) triang(filter_order+1) flattopwin(filter_order+1) nuttallwin(filter_order+1)};
window_type = cell2mat(window_type(window));

% Coeficientes con precision infinita
b = fir1(filter_order, fc, filter_type, window_type);
[H,w] = freqz(b,1);

% Cuantificacion
q = quantizer('fixed','round','saturate',[wordlength_coeff fractionallength_coeff]);
b_q = quantize(q,b);
[H_q,w_q] = freqz(b_q,1);

% Figures
figure
subplot(2,1,1)
plot(w/2/pi*fs,20*log10(abs([H H_q])));
legend('Infinite precision', [num2str(wordlength_coeff) ' bits precision']);
xlabel('Frequency (Hz)')
ylabel('Module (dB)')
grid

subplot(2,1,2)
plot(w/2/pi*fs,unwrap(angle([H H_q])));
legend('Infinite precision', [num2str(wordlength_coeff) ' bits precision']);
xlabel('Frequency (Hz)')
ylabel('Phase (rad)')
grid


%% Filtrado de la senyal - Filtro secuencial, con coeficientes simetrico y de orden par %%
% Filtro con truncado interno
sr = zeros(1,length(b));
response_q = zeros(1,length(signal));
for i = 1:length(signal_q)
    % shift register
    for j = length(b)-1:-1:1
        sr(j+1) = sr(j);
    end
    sr(1) = signal_q(i);
    
    % sequential filter
    acc = 0;
    for k = 1:ceil(length(b)/2)
        if k == ceil(length(b)/2)
            sum = sr(k);
        else
            sum = sr(k) + sr(length(b)+1-k);
        end
        q = quantizer('fixed','round','saturate',[wordlength_sum fractionallength_sum]);
        sum_q = quantize(q,sum);
        
        mult = b_q(k)*sum_q;
        q = quantizer('fixed','round','saturate',[wordlength_mult fractionallength_mult]);
        mult_q = quantize(q,mult);
        
        q = quantizer('fixed','round','saturate',[wordlength_acc fractionallength_acc]);
        acc_q = quantize(q,acc);
        acc = acc_q + mult_q;
    end
    q = quantizer('fixed','round','saturate',[wordlength_output fractionallength_output]);
    response_q(i) = quantize(q,acc);
end

% Filtro con crecimiento natural dentro del filtro pero con el estimulo y
% la respuesta cuantificadas
response = filter(quantize(q,b),1,signal_q);


%% Figuras %%
stimulispectrum = abs(fft(signal_q));
response_q_spectrum = abs(fft(response_q));
response_spectrum = abs(fft(response));

figure
subplot(2,2,1)
plot(Ts,signal_q)
title("Estimulo en el tiempo")

subplot(2,2,2)
plot(0:fs/length(stimulispectrum):fs-fs/length(stimulispectrum),stimulispectrum)
title("Estimulo en frecuencia")

subplot(2,2,3)
plot(Ts,response_q)
hold on
plot(Ts,response)
hold off
legend('Respuesta filtro cuantificado', 'Respuesta filtro con crecimiento natural')
title("Respuesta del filtro al estimulo en el tiempo")

subplot(2,2,4)
plot(0:fs/length(response_q_spectrum):fs-fs/length(response_q_spectrum),response_q_spectrum)
hold on
plot(0:fs/length(response_spectrum):fs-fs/length(response_spectrum),response_spectrum)
hold off
legend('Espectro filtro cuantificado', 'Espectro filtro con crecimiento natural')
title("Respuesta en frecuencia del filtro al estimulo")


%% Exportar datos cuantificados %%
% Coeficientes del filtro en formato binario y float
coeficientes = fopen('coefficients_b.txt','w');
q = quantizer('fixed','round','saturate',[wordlength_coeff fractionallength_coeff]);
for i = 1:length(b)
    fprintf(coeficientes,[num2bin(q,b(i)) '\n']);
end

coeficientes = fopen('coefficients_f.txt','w');
for i = 1:length(b)
    fprintf(coeficientes,[num2str(b(i)) '\n']);
end

% Estimulo en formato binario y float
estimulo = fopen('stimuli_b.txt','w');
q = quantizer('fixed','round','saturate',[wordlength_input fractionallength_input]);
for i = 1:length(signal)
    fprintf(estimulo,[num2bin(q,signal(i)) '\n']);
end

estimulo = fopen('stimuli_f.txt','w');
for i = 1:length(signal)
    fprintf(estimulo,[num2str(signal(i)) '\n']);
end

% Respuesta en formato binario y float
respuesta = fopen('response_b.txt','w');
q = quantizer('fixed','round','saturate',[wordlength_output fractionallength_output]);
for i = 1:length(response_q)
    fprintf(respuesta,[num2bin(q,response_q(i)) '\n']);
end

respuesta = fopen('response_f.txt','w');
for i = 1:length(response_q)
    fprintf(respuesta,[num2str(response_q(i)) '\n']);
end

File: Python/SFIRTI.py
"""
    Engineer: Yohan Curbelo Angles
    
    Module Name: SFIRTI (Sequential FIR Filter Type I)

    Description: 
    1. Customazible and optimized Sequential FIR Filter.
    2. The order of the filter has to be even (odd number of coefficients) 
    and its coefficients symmetric.
    3. Customizable: The widths of input, output, and internal signals are 
    configurable, making natual growth and truncation of data possible. 
    4. Optimized: As filter coefficients are symmetric, only half of values 
    that are symmetric to the middle address of the shift register are summed.

    Additional notes:
    1. Stimuli is the result of adding 3 tones.
    2. Stimuli, filter response, and filter coefficientes are exported to a
    txt file.

    Descripcion:
    1. Filtro FIR secuencial personalizable.
    2. El orden del filtro tiene que ser par (cantidad impar de coeficientes)
    y sus coeficientes simetricos.
    3. Personalizable: El ancho de la senyal de entrada, salida, e internas son 
    configurables, haciendo posible el truncado y el crecimiento natural de los 
    datos internos.
    4. Optimizado: Como el filtro es simetrico solo se realiza una suma de los 
    valores que son simetricos respecto a la direccion central del registro de 
    desplazamiento.
    
    Notas adicionales:
    1. La senyal de estimulo es la superposicion de 3 tonos.
    2. Se exportan a txt la senyal de estimulo, la respuesta del filtro, y los
    valores de los coeficientes del filtro.
"""

import numpy as np
import my_functions as mf
import matplotlib.pyplot as plt
from scipy import signal
from scipy.fft import fft


# 1. CONFIGURABLE PARAMETERS
# Filter especification
fc1 = 2000                              # Cut-off frequency 1
fc2 = 3000                              # Cut-off frequency 2. Only for bandpass or bandstop filters
FILTER_ORDER = 50	                    # Filter order
TOTAL_COEFFICIENTS = FILTER_ORDER + 1   # Total of coefficientes is order + 1
filter_type = 'lowpass'                 # Options: lowpass; highpass; bandpass; bandstop
window_type = 'hann'                    # Options: hann; hamming; blackman; triang; flattop; nuttall
COEFFICIENTS_LENGTH = 10
COEFFICIENTS_FRACTIONAL_LENGTH = 9

# Data growth
INPUT_LENGTH = 12
INPUT_FRACTIONAL_LENGTH = 10
SUM_LENGTH = 12
SUM_FRACTIONAL_LENGTH = 9
MULT_LENGTH = 12
MULT_FRACTIONAL_LENGTH = 8
ACC_LENGTH = 18
ACC_FRACTIONAL_LENGTH = 8
OUTPUT_LENGTH = 12
OUTPUT_FRACTIONAL_LENGTH = 3

# Stimuli
pi = np.pi
A = 2                   # Amplitud of input signal 
f1 = 300
f2 = 1000
f3 = 2400
fsample = 8000          # Sampling frequency
t_sim = 0.1             # Simulation time (seconds)
Ts = np.arange(0, t_sim, 1/fsample)
stimuli = A/3*np.sin(2*pi*f1*Ts) + A/3*np.sin(2*pi*f2*Ts) + A/3*np.sin(2*pi*f3*Ts)

# 2. VALIDATION OF THE INPUTS
if FILTER_ORDER%2 != 0:
    raise Exception('Filter order has to be even')
elif filter_type != 'lowpass' and filter_type != 'highpass' and filter_type != 'bandpass' and filter_type != 'bandstop':
    raise Exception('Invalid filter type.')
elif filter_type == 'bandpass' or filter_type == 'bandstop':
    if fc2 <= fc1:
        raise Exception('Cut-off frequency has to be increasing') 
    if fsample/2 < fc2:
        raise Exception('Cutt-off frequency cannot be higher than Nyquist frequency') 
elif fsample/2 < fc1:
    raise Exception('Cutt-off frequency cannot be higher than Nyquist frequency') 
elif window_type != 'hann' and window_type != 'hamming' and window_type != 'blackman' and window_type != 'triang' and window_type != 'flattop' and window_type != 'nuttall':
    raise Exception('Invalid window type.')
elif COEFFICIENTS_FRACTIONAL_LENGTH >= COEFFICIENTS_LENGTH:
    raise Exception('Coefficients length has to be larger than its fractional length')
elif INPUT_FRACTIONAL_LENGTH >= INPUT_LENGTH:
    raise Exception('Input length has to be larger than its fractional length')
elif SUM_FRACTIONAL_LENGTH >= SUM_LENGTH:
    raise Exception('Sum length has to be larger than its fractional length')
elif MULT_FRACTIONAL_LENGTH >= MULT_LENGTH:
    raise Exception('Multiplication length has to be larger than its fractional length')
elif ACC_FRACTIONAL_LENGTH >= ACC_LENGTH:
    raise Exception('Accumulator length has to be larger than its fractional length')
elif OUTPUT_FRACTIONAL_LENGTH >= OUTPUT_LENGTH:
    raise Exception('Output length has to be larger than its fractional length')


# 3. FILTER DESIGN
# Set the cut-off frequency
if filter_type == 'bandpass' or filter_type == 'bandstop':
    fc = [fc1, fc2]
else:
    fc = fc1
    
# Coefficients     
num = signal.firwin(TOTAL_COEFFICIENTS, fc, window=window_type, pass_zero=filter_type, fs=fsample)

# Quantization
num_q = mf.quantize(num, COEFFICIENTS_LENGTH, COEFFICIENTS_FRACTIONAL_LENGTH)
stimuli_q = mf.quantize(stimuli, INPUT_LENGTH, INPUT_FRACTIONAL_LENGTH)

# Optimized filter
sr = np.zeros(TOTAL_COEFFICIENTS)             # There are FILTER_ORDER + 1 coefficients
response = np.zeros(len(stimuli_q))
response_q = np.zeros(len(stimuli_q))
for i in range(len(stimuli_q)):
    # Shift register stage
    for j in range(TOTAL_COEFFICIENTS - 1):
        sr[TOTAL_COEFFICIENTS - 1 - j] = sr[TOTAL_COEFFICIENTS - j - 2]
    sr[0] = stimuli_q[i]
    
    # Filter stage
    acc = 0
    acc_q = 0
    for k in range(int(np.ceil(TOTAL_COEFFICIENTS/2))):
        # Sum
        if k != int(TOTAL_COEFFICIENTS/2):
            sum = sr[k] + sr[TOTAL_COEFFICIENTS - 1 - k]
        else:
            sum = sr[int(TOTAL_COEFFICIENTS/2)]
        sum_q = mf.quantize(sum, SUM_LENGTH, SUM_FRACTIONAL_LENGTH) 
            
        # Multiplication
        mult = sum_q * num_q[k]
        mult_q = mf.quantize(mult, MULT_LENGTH, MULT_FRACTIONAL_LENGTH)
        
        # Accumulator
        acc_q = mf.quantize(acc, ACC_LENGTH, ACC_FRACTIONAL_LENGTH)
        acc = acc_q + mult_q
        
    response_q[i] = mf.quantize(acc, OUTPUT_LENGTH, OUTPUT_FRACTIONAL_LENGTH)

# 4. Figures
stimuli_spectrum = abs(fft(stimuli_q))
response_spectrum = abs(fft(response_q))

time_axis = np.arange(0, t_sim, t_sim/len(stimuli_q))
freq_axis = np.arange(0, fsample, fsample/len(response_q))

# Filter plot
w, h = signal.freqz(num_q)
fig, axe = plt.subplots()
axe.set_title('Digital filter frequency response')
axe.plot(w/2/pi*fsample, 20 * np.log10(abs(h)), 'b')
axe.set_ylabel('Amplitude [dB]', color='b')
axe.set_xlabel('Frequency [Hz]')
axe2 = axe.twinx()
angles = np.unwrap(np.angle(h))
axe2.plot(w/2/pi*fsample, angles, 'g')
axe2.set_ylabel('Angle [radians]', color='g')
axe2.grid()
axe2.axis('tight')

# Stimuli and response plot
fig, axe = plt.subplots(2, 2)
axe[0,0].plot(time_axis, stimuli_q)
axe[0,0].set_title('Stimuli (time domain')
axe[1,0].plot(freq_axis, stimuli_spectrum)
axe[0,0].set_title('Stimuli (frequency domain')
axe[0,1].plot(time_axis, response_q)
axe[0,0].set_title('Response (time domain')
axe[1,1].plot(freq_axis, response_spectrum)
axe[0,0].set_title('Response (frequency domain')
fig.suptitle('Filter behavior')

# Plot figures
plt.show()

# 5. Export data in float and binary format
# Float files
with open('Python\Stimuli_f.txt','w') as file:
    for line in stimuli_q:
        file.write(str(line))
        file.write('\n')
        
with open('Python\Coefficients_f.txt','w') as file:
    for line in num_q:
        file.write(str(line))
        file.write('\n')
        
with open('Python\Response_f.txt','w') as file:
    for line in response_q:
        file.write(str(line))
        file.write('\n')       
        
# Binary files
with open('Python\Stimuli_b.txt','w') as file:
    # Convert to binary representation
    stimuli_qb = ["" for i in range(len(stimuli_q))]        
    for i in range(len(stimuli_q)):        
        stimuli_qb[i] = mf.float_to_bin(stimuli_q[i], INPUT_LENGTH, INPUT_FRACTIONAL_LENGTH)
    # Write txt file
    for line in stimuli_qb:
        file.write(str(line))
        file.write('\n')

with open('Python\Coefficients_b.txt','w') as file:
    # Convert to binary representation
    num_qb = ["" for i in range(len(num_q))]           
    for i in range(len(num_q)):   
        num_qb[i] = mf.float_to_bin(num_q[i], COEFFICIENTS_LENGTH, COEFFICIENTS_FRACTIONAL_LENGTH)
    # Write txt file
    for line in num_qb:
        file.write(str(line))
        file.write('\n')
      
with open('Python\Response_b.txt','w') as file:
    # Convert to binary representation
    response_qb = ["" for i in range(len(response_q))]           
    for i in range(len(response_q)):                         
        response_qb[i] = mf.float_to_bin(response_q[i], OUTPUT_LENGTH, OUTPUT_FRACTIONAL_LENGTH)     
    # Write txt file
    for line in response_qb:
        file.write(str(line))
        file.write('\n')